lldb/source/Core/Progress.cpp

```diff
@@ -66,3 +66,47 @@ void Progress::ReportProgress() {
                              m_debugger_id);
   }
 }
+
+void ProgressManager::Initialize() {
+  lldbassert(!InstanceImpl() && "A progress report manager already exists.");
+  InstanceImpl().emplace();
+}
+
+void ProgressManager::Terminate() {
+  lldbassert(InstanceImpl() &&
+             "A progress report manager has already been terminated.");
+  InstanceImpl().reset();
+}
+
+std::optional<ProgressManager> &ProgressManager::InstanceImpl() {
+  static std::optional<ProgressManager> g_progress_manager;
+  return g_progress_manager;
+}
+
+ProgressManager::ProgressManager() : m_progress_map() {}
+
+ProgressManager::~ProgressManager() {}
+
+ProgressManager &ProgressManager::Instance() { return *InstanceImpl(); }
```

Comment:
We have other subsystems that work exactly like what Chelsea did here, such as the Diagnostics and the FileSystem. They rely on the requirement that you need to call `:: Initialize ` before doing anything with it.

Reply:
> We have other subsystems that work exactly like what Chelsea did here, such as the Diagnostics and the FileSystem. They rely on the requirement that you need to call `::Initialize ` before doing anything with it. It's true that this will crash if you forget to `Initialize`/`Terminate` but that's true for the `Debugger` too. This is an establishes pattern and has the advantage that it can be tested in unittest by initializing and terminating the subsystem between tests.

This is fine as long as you always check the optional then and we aren't doing that. So this can and will crash LLDB if extra threads call these APIs which is never good. So the API, in order to be safe, should be:
```
std::optional<ProgressManager &> ProgressManager::InstanceImpl() 
```
And then every user needs to check this optional. Otherwise, this:
```
ProgressManager &ProgressManager::Instance() { return *InstanceImpl(); }
```
Will crash if Terminate has been called right? the `std:optional::operator *`docs say "The behavior is undefined if *this does not contain a value."

Reply:
This can cause shut down crashes in LLDB where extra threads call a progress API.

Reply:
Another reason for using the `std::once` is if no one tries to listen to the eBroadcastBitProgressCategory, then nothing will need to be created. If not we are always making an instance. Granted this object isn't very expensive to create, but it is extra work that might not be needed.

Reply:
I also want to say that I don't mind the need to call `ProgressManager::Initialize()` before using parts of this class. I am mostly looking out for shut down crashes of the the LLDB library where the main thread exits first while other threads are not all terminated, which does happen. All of the places we have the comment:
```
// NOTE: intentional leak to avoid issues with C++ destructor chain
```
Was from tracking down shut down crashers in LLDB.

Reply:
To my understanding this behaviour of the Debugger caused an issue in the unit test suite when I added the test for progress reporting since it tried to initialize the debugger after another test had already done so. Could we just place the `once_flag` in `InstanceImpl()` but keep `Initialize/Terminate` to keep the pattern we already have without having the user always check the optional?

Reply:
Yeah, I understand why we might want to leak the vector and do in other places. But do you think we should just do that unconditionally, no matter whether it's a problem or not? The other subsystems I mention use the same approach and they don't crash in production even though they don't check the optional before dereferencing it (treating it as a precondition). 

I guess the point I'm trying to make is let's not leak it proactively, but use the stricter approach to force a crash/assert in debug builds and tests, and try to avoid these issues, rather than work around them. If we find that it does end up crashing because someone emits a progress report when we're shutting down LLDB **and** we've investigated the bug and think the functionality is important, I'm totally on board with leaking it. Does that sound like a reasonable compromise?

Reply:
It is fine to start with this and switch over eventually if we run into problems. Not sure what percentage of people will listen to the category progresses vs the individual ones. So it will affect only the people using the category ones.

Reply:
But wether you leak it or not it won't affect the program since it is exiting already when this gets cleaned up if it is left as is above. So up to you guys.

Reply:
So we'll keep the current functionality (using `Initialize/Terminate`, and having the user check the optional instead of doing it in the class) if I'm understanding correctly?

Reply:
Works for me as long as all clients check the optional and use it correctly. It allows for multi-thread safety so that is good. Not sure if std::optional is actually threadsafe though, but the timing required to make this fail would be pretty low percentage. If it were me I would make it bullet proof as just leak the map as when we tear down the process, any allocated memory will just go poof anyway and it will actually speed up the process teardown process as no work will need to be done to clear this map, though it should really be empty anyway.

Reply:
Given that this patch follows a pattern established by the other subsystems, and those subsystems don't currently show any major issues in production with their initialization and termination lifecycle I think we can follow the current subsystem behaviour for this patch and discussions about the subsystems themselves can take place in a separate discussion.